src/engine/PreconditionAction.h

```diff
@@ -0,0 +1,84 @@
+//   Copyright 2025, University of Freiburg,
+//   Chair of Algorithms and Data Structures.
+//   Author: Robin Textor-Falconi <textorr@informatik.uni-freiburg.de>
+
+#ifndef PRECONDITIONACTION_H
+#define PRECONDITIONACTION_H
+
+#include <memory>
+#include <variant>
+
+#include "backports/concepts.h"
+#include "util/Exception.h"
+#include "util/TypeTraits.h"
+
+// forward declaration needed to break dependencies
+class QueryExecutionTree;
+
+// Represent how a precondition of an operation can be satisfied.
+// `isImplicitlySatisfied()` will return true if the `Operation` in question is
+// already naturally fulfilling the requested requirement.
+// `mustBeSatisfiedExternally()` will return true if the `Operation` cannot
+// fulfil the requirement on its own and needs external support.
+// These two cases are mutually exclusive. If both options are false, the
+// operation created a query execution tree that represents a modification of
+// itself that can fulfil the requirement.
+class PreconditionAction {
```

Comment:
Consider moving this into a namespace
(global namespace cluttering is bad). 
maybe the `qlever` namespace or something.